src/distributed_graph_generator.cc

```diff
@@ -315,27 +315,33 @@ void distributed_graph_generator::generate_distributed_commands(const task& tsk)
 							buffer_state.local_last_writer.update_region(missing_parts, {ap_cmd->get_cid(), true /* is_replicated */});
 						}
 					} else if(!is_pending_reduction) {
+						// We generate separate push command for each last writer command for now, possibly even multiple for partially already-replicated data.
+						// TODO: Can and/or should we consolidate?
 						const auto local_sources = buffer_state.local_last_writer.get_region_values(req);
 						for(const auto& [local_box, wcs] : local_sources) {
 							if(!wcs.is_fresh() || wcs.is_replicated()) { continue; }
 
-							// Check if we've already pushed this box
-							const auto replicated_boxes = buffer_state.replicated_regions.get_region_values(local_box);
-							for(const auto& [replicated_box, nodes] : replicated_boxes) {
+							// Make sure we don't push anything we've already pushed to this node before
+							box_vector<3> non_replicated_boxes;
+							for(const auto& [replicated_box, nodes] : buffer_state.replicated_regions.get_region_values(local_box)) {
 								if(nodes.test(nid)) continue;
+								non_replicated_boxes.push_back(replicated_box);
+							}
 
-								// Generate separate push command for each last writer command for now,
-								// possibly even multiple for partially already-replicated data.
-								// TODO: Can and/or should we consolidate?
-								auto* const push_cmd = create_command<push_command>(bid, 0, nid, trid, replicated_box.get_subrange());
+							// Merge all connected boxes to determine final set of pushes
+							const auto push_boxes = region<3>(std::move(non_replicated_boxes));
+							for(auto& push_box : push_boxes.get_boxes()) {
```

Comment:
```suggestion
							const auto push_boxes = region<3>(std::move(non_replicated_boxes)).into_boxes();
							for(auto& push_box : push_boxes.into_boxes()) {
```
I just remembered that `region::into_boxes() &&` exists, that way the variable name also makes more sense.

Reply:
I need the region again below the loop for updating `buffer_state.replicated_regions`. Renamed to `push_region`!